src/concepts/mods.md

```diff
@@ -1 +1,506 @@
+operands to the operation to be performed. While we would like an implementation
+of this signature to also display the behavior outlined in the comments,
+realistically there is no way to verify that this is true (why?), so the best that we
+can do is ensure that an implementation provides functions with the correct
+type.
+
+Note also that this syntax is slightly different than that you have seen before
+- we use `val` as if we were going to produce a val binding (with an associated
+  type annotation), however we do not actually give it any definition. In truth,
+you _cannot_ define a value in the signature itself, since it's impossible in
+general to check if two values are the same. You can, however, define a `type`
+within a signature (or leave it definitionless, as we will see later in the
+chapter).
+
+An implementation of `MOD_ARITH` (called a _structure_, or _module_) might look as follows:
+```sml
+structure ModArith : MOD_ARITH =
+struct
+    fun mod_add n x y = (x + y) mod n
+
+    fun mod_sub n x y = mod_add n x (~y)
+
+    fun mod_times n x y =
+        case Int.compare (x, 0) of
+            LESS => mod_add n (mod_times n (x + 1) y) (~y)
+          | EQUAL => 0
+          | GREATER => mod_add n (mod_times n (x - 1) y) y
+end
+```
+
+**NOTE**: By convention, structure names are usually capitalized, and signature
+names are usually all-capitals. Additionally, it is totally fine to declare a
+structure's values with `fun` instead of `val`, even if it says `val` in the
+signature, as `fun` is just shorthand for producing a function value binding
+anyways.
+
+We will go over the precise meaning of the usage of the colon symbol in the
+above structure later in the chapter. For now, the meaning of this code snippet
+is to produce a structure _ascribing_ to the `MOD_ARITH` signature, or in other
+words, implementing the `MOD_ARITH` specification. We see that it contains three
+declarations, those being the functions declared in the signature itself. A key
+note is that ascribing to a signature is akin to a contract - any structure
+ascribing a signature _must_ implement all the requisite components described by
+the signature, or it will fail to ascribe, and result in a compile-time error
+(similar to an ill-typed program).
+
+It is not, however, the case that a structure cannot provide _more_ information
+than is strictly necessitated by the signature. Additional helper functions and
+value bindings can be freely instantiated within a structure without affecting
+ascription. Thus, honoring a contract only entails satisfyng the terms agreed to
+in the signature, without comment on going over. We will explore this idea more
+later in the chapter when we discuss information hiding.
+
+Using structures should be something that you are already familiar with - you do
+it every time that you invoke `Int.compare`. To use the fields of a structure,
+you access them using the dot operator, following the name of the structure in
```

Comment:
maybe "you write the structure name, then a dot, and then the field of the structure"